packages/plugin-pack/sources/index.ts

```diff
@@ -58,6 +58,9 @@ const beforeWorkspacePacking = (workspace: Workspace, rawManifest: any) => {
         // For workspace:path/to/workspace and workspace:* we look up the workspace version
         if (structUtils.areDescriptorsEqual(descriptor, matchingWorkspace.anchoredDescriptor) || range.selector === `*`)
           versionToWrite = matchingWorkspace.manifest.version ?? `0.0.0`;
+        // For workspace:~ and workspace:^ we add selector in front of workspace version
+        else if (range.selector === `~` || range.selector === `^`)
+          versionToWrite =  `${range.selector}${matchingWorkspace.manifest.version}`;
```

Comment:
(Same fallback as in the `if` above)
```suggestion
          versionToWrite =  `${range.selector}${matchingWorkspace.manifest.version ?? `0.0.0`}`;
```